UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/TensorApplier.cs

```diff
@@ -0,0 +1,143 @@
+﻿using UnityEngine.MachineLearning.InferenceEngine;
+using System.Collections.Generic;
+using UnityEngine.MachineLearning.InferenceEngine.Util;
+
+namespace MLAgents.InferenceBrain
+{
+    public interface TensorApplier
+    {
+        void Execute(Tensor tensor, Dictionary<Agent, AgentInfo> agentInfo);
+    }
+
+    /// <summary>
+    /// The Applier for the Continuous Action output tensor.
+    /// </summary>
+    public class ContinuousActionOutputApplier : TensorApplier
+    {
+        public void Execute(Tensor tensor, Dictionary<Agent, AgentInfo> agentInfo)
+        {
+            var tensorDataAction = tensor.Data as float[,];
+            var actionSize = tensor.Shape[1];
+            var agentIndex = 0;
+            foreach (var agent in agentInfo.Keys)
+            {
+                var a = new float[actionSize];
+                for (var j = 0; j < actionSize; j++)
+                {
+                    a[j] = tensorDataAction[agentIndex, j];
+                }
+
+                agent.UpdateVectorAction(a);
+                agentIndex++;
+            }
+        }
+    }
+
+    /// <summary>
+    /// The Applier for the Discrete Action output tensor. Uses multinomial to sample discrete
+    /// actions from the logits contained in the tensor.
+    /// </summary>
+    public class DiscreteActionOutputApplier : TensorApplier
+    {
+        private int[] _actionSize;
+        private Multinomial _multinomial;
+        
+        public DiscreteActionOutputApplier(int[] actionSize, int seed)
+        {
+            _actionSize = actionSize;
+            _multinomial = new Multinomial(seed);
+        }
+        
+        public void Execute(Tensor tensor, Dictionary<Agent, AgentInfo> agentInfo)
+        {
+            var tensorDataProbabilities = tensor.Data as float[,];
+            var batchSize = agentInfo.Keys.Count;
+            var actions = new float[batchSize, _actionSize.Length];
+            var startActionIndices = Utilities.CumSum(_actionSize);
+            for (var actionIndex=0; actionIndex < _actionSize.Length; actionIndex++)
+            {
+                var nBranchAction = startActionIndices[actionIndex + 1] -
```

Comment:
Isn't nBranchAction also equal to `_actionSize[actionIndex]`

more broadly, if `b = cumsum(a)` then `a[x] = b[x+1] - b[x]` (but admittedly, i may be missing something)